README.md
# QuickCSF

A fast, adaptive approach to estimating contrast sensitivity function parameters.

This implementation is based on:

> Lesmes, L. A., Lu, Z. L., Baek, J., & Albright, T. D. (2010). Bayesian adaptive estimation of the contrast sensitivity function: The quick CSF method. *Journal of vision*, 10(3), 17-17.

Special thanks to Dr. Tianshi Lu at Wichita State University for providing a MATLAB implementation of the fundamental algorithm and Dr. Rui Ni at Wichita State University for the motivation.

## Dependencies

Dependencies are installed via the `conda env` command. They include ---

* `numpy` - numerical calculations
* `qtpy` - mystical py-glue
* `pyqt` - python bindings, right now at v 5.9 in Anaconda, 5.13 in the wild.

<!-- 
Requires:

* `numpy`
* `qtpy`
* Qt bindings (via `PySide2`, `PyQt5`, `PySide`, or `PyQt`)

Optional (for simulation visuals):

* `matplotlib` -->

## Installation

<!-- 
~~~~bash
pip3 install -e .
~~~~ -->

To install into a environment using Anaconda, use the description found in the `conda` directory.

When using Anaconda, you should **not** run the supplied `setup.py`, but instead, use the `conda env` command. See the `README.md` in the `conda/` directory for information.

## Usage

### Measuring CSF

Ensure you are using the `qcsf` environment in Anaconda, using:

~~~bash
conda activate qcsf
~~~

Run:

~~~bash
python -m QuickCSF.app
~~~

A settings dialog will appear; session ID and viewing distance are required. Arguments can also be specified on the command line. Use the `--help` flag to see all options:

~~~bash
python -m QuickCSF.app --help
~~~

### Simulate and visualize an evaluation

Run:

~~~bash
python -m QuickCSF.simulate
~~~

A settings dialog will appear; the number of trials is required. Arguments can also be specified on the command line. use the `--help` flag to see all options:

~~~bash
python -m QuickCSF.simulate --help
~~~


CHANGELOG.md
# Change Log

QuickCSF

<!--
## Guiding Principles

- Changelogs are for humans, not machines. 
- There should be an entry for every single version.
- The same types of changes should be grouped.
- Versions and sections should be linkable.
- The latest version comes first.
- The release date of each versions is displayed.
- Mention whether you follow Semantic Versioning.

## Types of changes

- `Added` for new features.
- `Changed` for changes in existing functionality.
- `Deprecated` for soon-to-be removed features.
- `Removed` for now removed features.
- `Fixed` for any bug fixes.
- `Security` in case of vulnerabilities. 
-->

## [Unreleased changes]

## [1.0.1] - 2019-06-19

Seems to actually work cross-platform.

### Added

- `conda` environment description
- instructions in `README.md`
- my versioning and building code.

### Changed

- Made markdown formatting 'compliant' for `README`
- Tweaked some formatting for PEP8 compatibility (FP=OCD) _but_ I did keep tab-spacing.
- added my vscode configurations mainly for x-platform-ism (e.g., removed from `.gitignore`)

### `Deprecated`

- Warning and exit in `setup.py`.

### Fixed

- `setPixmap(None)` can't be `None`
- `argv` for widget
- temporary regression of `screenAt` to support Anaconda's 'mere' support of Qt 5.9
- `argv` for screen.py test code

## [0.0.1] - 2019-05-20

Forked from `vpclab/QuickCSF`


QuickCSF/ui.py
# -*- coding: utf-8 -*
'''User interaction (display and input)'''

import sys, traceback
import logging
import math, time

import numpy

from qtpy import QtCore, QtGui, QtWidgets, QtMultimedia
import argparseqt.gui

from . import assets

logger = logging.getLogger(__name__)

DEFAULT_INSTRUCTIONS = '''For this test, you will be presented with two options - one will be blank, and the other will be a striped circle.\n\n
A tone will play when each option is displayed. After both tones, you will need to select which option contained the striped circle.\n\n
If the striped circle appeared during the FIRST tone, press [ ← LEFT ].\n
If the striped circle appeared during the SECOND tone, press [ RIGHT → ].\n\n
Throughout the test, keep your gaze fixated on the circled-dot at the center of the screen.\n\n
If you are uncertain, make a guess.\n\n\nPress [ SPACEBAR ] to start.'''

class QuickCSFWindow(QtWidgets.QMainWindow):
	'''The main window for QuickCSF.app'''

	participantReady = QtCore.Signal()
	participantResponse = QtCore.Signal(object)

	def __init__(self, instructions=None, parent=None):
		super().__init__(parent)
		self.displayWidget = QtWidgets.QLabel(self)
		self.displayWidget.setAlignment(QtCore.Qt.AlignHCenter|QtCore.Qt.AlignVCenter)
		self.displayWidget.setWordWrap(True)
		self.displayWidget.setMargin(100)
		self.displayWidget.setStyleSheet(
			'''
				background: rgb(127, 127, 127);
				color: #bbb;
				font-size: 28pt;
			'''
		)

		self.instructionsText = instructions if instructions is not None else DEFAULT_INSTRUCTIONS
		
		self.breakText = 'Good job - it\'s now time for a break!\n\nWhen you are ready to continue, press the [ SPACEBAR ].'
		self.readyText = 'Ready?'
		self.responseText = '?'
		self.finishedText = 'All done!'

		self.setCentralWidget(self.displayWidget)
		self.sounds = {
			'tone': QtMultimedia.QSound(assets.locate('tone.wav')),
			'good': QtMultimedia.QSound(assets.locate('good.wav')),
			'bad': QtMultimedia.QSound(assets.locate('bad.wav')),
		}

	def showInstructions(self):
		self.displayWidget.setText(self.instructionsText)

	def showReadyPrompt(self):
		self.displayWidget.setText(self.readyText)

	def showFixationCross(self):
		self.displayWidget.setText('+')

	def showStimulus(self, stimulus):
		self.displayWidget.setPixmap(QtGui.QPixmap.fromImage(stimulus))
		self.sounds['tone'].play()

	def showNonStimulus(self):
		self.showBlank()
		self.sounds['tone'].play()

	def showMask(self):
		self.displayWidget.setText('')

	def showBlank(self):
		self.displayWidget.setText('')
		self.displayWidget.setPixmap(QtGui.QPixmap())

	def giveFeedback(self, good):
		if good:
			self.displayWidget.setText('Good!')
			self.sounds['good'].play()
		else:
			self.displayWidget.setText('Wrong')
			self.sounds['bad'].play()

	def showResponsePrompt(self):
		self.displayWidget.setText(self.responseText)

	def showBreak(self):
		self.displayWidget.setText(self.breakText)

	def showFinished(self, results):
		outputDisplay = self.finishedText + '\n'
		for key,value in results.items():
			outputDisplay += f'\n{key} = {value:.4f}'
				
		self.displayWidget.setText(outputDisplay)

	def keyReleaseEvent(self, event):
		logger.debug(f'Key released {event.key()}')
		if event.key() == QtCore.Qt.Key_Space:
			self.participantReady.emit()
		elif event.key() in (QtCore.Qt.Key_4, QtCore.Qt.Key_Left):
			self.participantResponse.emit(True)
		elif event.key() in (QtCore.Qt.Key_6, QtCore.Qt.Key_Right):
			self.participantResponse.emit(False)

	def onNewState(self, stateName, data):
		logger.debug(f'New state: {stateName} [{data}]')

		if stateName == 'INSTRUCTIONS':
			self.showInstructions()
		elif stateName == 'BREAKING':
			self.showBreak()
		elif stateName == 'WAIT_FOR_READY':
			self.showReadyPrompt()
		elif 'FIXATION' in stateName:
			self.showFixationCross()
		elif '_BLANK' in stateName:
			self.showBlank()
		elif stateName == 'SHOW_STIMULUS_1':
			if data.stimulusOnFirst:
				self.showStimulus(data.stimulus)
			else:
				self.showNonStimulus()
		elif stateName == 'SHOW_MASK_1':
			self.showMask()
		elif stateName == 'SHOW_STIMULUS_2':
			if not data.stimulusOnFirst:
				self.showStimulus(data.stimulus)
			else:
				self.showNonStimulus()
		elif stateName == 'SHOW_MASK_2':
			self.showMask()
		elif stateName == 'WAIT_FOR_RESPONSE':
			self.showResponsePrompt()
		elif stateName == 'FEEDBACK':
			self.giveFeedback(data.correct)
		elif stateName == 'FINISHED':
			self.showFinished(data)

def getSettings(parser, settings, requiredFields=[]):
	'''Display a GUI to collect experiment settings'''
	dialog = argparseqt.gui.ArgDialog(parser)
	dialog.setValues(settings)
	dialog.exec_()
	if dialog.result() == QtWidgets.QDialog.Accepted:
		settings = dialog.getValues()
		
		for field in requiredFields:
			if settings[field] == None:
				QtWidgets.QMessageBox.critical(
					None,
					'Missing information',
					'You must specify the following fields: ' + ','.join(requiredFields)
				)
				return None
	else:
		return None

	return settings

def exception_handler(excType, exc, tb, extraDetails=None, parentWindow=None):
	if issubclass(excType, KeyboardInterrupt):
		sys.__excepthook__(excType, exc, tb)
		return

	stack = traceback.format_tb(tb)

	details = '%s: %s\n%s' % (excType.__name__, exc, ''.join(stack))
	print('UNHANDLED EXCEPTION! ' + details, file=sys.stderr)

	dialog = QtWidgets.QMessageBox(parentWindow)
	dialog.setWindowTitle('QuickCSF - Application Error')
	dialog.setText('Something went wrong and we were not prepared for it :(\n\nThe application will now exit, but some details may be available below.')

	if extraDetails is not None:
		details = extraDetails + '\n\n' + details

	dialog.setDetailedText(details)
	dialog.setModal(True)
	dialog.exec_()
	QtWidgets.QApplication.exit()

def popupUncaughtExceptions(extraDetails=None, parent=None):
	sys.excepthook = lambda excType, exc, tb: exception_handler(excType, exc, tb, extraDetails, parent)


QuickCSF/app.py
# -*- coding: utf-8 -*
'''An simple QuickCSF app to measure full contrast sensitivity function

	Executes a series of trials using 2AFC to measure CSF. Results are saved to a .CSF file.

	Example:
		$ python3 -m QuickCSF.app --help
		$ python3 -m QuickCSF.app -d 750 -s participant001
		$ python3 -m QuickCSF.app -d 750 --controller.trialsPerBlock 50 --controller.blockCount 2
'''

import logging

import argparse
import sys, pathlib, csv
import functools
from datetime import datetime

from qtpy import QtWidgets, QtCore
import argparseqt.groupingTools

from . import ui
from . import CSFController
from . import QuickCSF
from . import StimulusGenerators
from . import screens

logger = logging.getLogger('QuickCSF.app')

app = QtWidgets.QApplication(sys.argv)
app.setApplicationName('QuickCSF')
mainWindow = None
settings = None

def _onFinished(results):
	outputFile = pathlib.Path(settings['outputFile'])
	logger.debug('Writing output file: ' + str(outputFile.resolve()))

	fileExists = outputFile.exists()
	with outputFile.open('a') as csvFile:
		record = {
			'SessionID': settings['sessionID'],
			'Timestamp': datetime.today().strftime('%Y-%m-%d %H:%M:%S'),
			**results
		}

		writer = csv.DictWriter(csvFile, fieldnames=record.keys())
		if not fileExists:
			writer.writeheader()

		writer.writerow(record)

def _start():
	global mainWindow, settings

	def onStateTransition(state, data):
		if state == 'FINISHED':
			_onFinished(data)

	logger.debug('Showing main window')

	if settings['instructionsFile'] is not None and settings['instructionsFile'] != '':
		with open(settings['instructionsFile']) as instructionsFile:
			instructions = instructionsFile.read()
	else:
		instructions = None

	mainWindow = ui.QuickCSFWindow(instructions)

	degreesToPixels = functools.partial(screens.degreesToPixels, distance_mm=settings['distance_mm'])

	stimGenerator = StimulusGenerators.QuickCSFGenerator(degreesToPixels=degreesToPixels, **settings['Stimuli'])
	controller = CSFController.Controller_2AFC(stimGenerator, **settings['Controller'])

	mainWindow.participantReady.connect(controller.onParticipantReady)
	mainWindow.participantResponse.connect(controller.onParticipantResponse)

	controller.stateTransition.connect(mainWindow.onNewState)
	controller.stateTransition.connect(onStateTransition)

	QtCore.QTimer.singleShot(0, controller.start)
	mainWindow.showFullScreen()

def run(configuredSettings=None):
	'''Start the QuickCSF app'''
	global settings
	
	settings = configuredSettings

	ui.popupUncaughtExceptions()
	QtCore.QTimer.singleShot(0, lambda: _start())
	app.exec_()
	logger.info('App exited')

def getSettings():
	'''Parse command line arguments for the QuickCSF app. Will show a UI promptif necessary information is missing'''

	parser = argparse.ArgumentParser()
	parser.add_argument('-sid', '--sessionID', default=None, help='A unique string to identify this observer/session')
	parser.add_argument('-d', '--distance_mm', type=float, default=None, help='Distance (mm) from the display to the observer')
	parser.add_argument('--outputFile', default='data/QuickCSF-results.csv', help='The path/file to save results into')
	parser.add_argument('--instructionsFile', default=None, help='A plaintext file containing the instructions. If unspecified, default instructions will be displayed')

	controllerSettings = parser.add_argument_group('Controller')
	controllerSettings.add_argument('--trialsPerBlock', type=int, default=25, help='Number of trials in each block')
	controllerSettings.add_argument('--blockCount', type=int, default=4, help='Number of blocks')

	controllerSettings.add_argument('--fixationDuration', type=float, default=.25, help='How long (seconds) the fixation stimulus is displayed')
	controllerSettings.add_argument('--stimulusDuration', type=float, default=.1, help='How long (seconds) the stimulus is displayed')
	controllerSettings.add_argument('--maskDuration', type=float, default=.1, help='How long (seconds) the stimulus mask is displayed')
	controllerSettings.add_argument('--interStimulusInterval', type=float, default=.1, help='How long (seconds) a blank is displayed between stimuli')
	controllerSettings.add_argument('--feedbackDuration', type=float, default=.5, help='How long (seconds) feedback is displayed')

	controllerSettings.add_argument('--waitForReady', default=False, action='store_true', help='Wait for the participant to indicate they are ready for the next trial')

	stimulusSettings = parser.add_argument_group('Stimuli')
	stimulusSettings.add_argument('-minc', '--minContrast', type=float, default=.01, help='The lowest contrast value to measure (0.0-1.0)')
	stimulusSettings.add_argument('-maxc', '--maxContrast', type=float, default=1.0, help='The highest contrast value to measure (0.0-1.0)')
	stimulusSettings.add_argument('-cr', '--contrastResolution', type=int, default=24, help='The number of contrast steps')

	stimulusSettings.add_argument('-minf', '--minFrequency', type=float, default=0.2, help='The lowest frequency value to measure (cycles per degree)')
	stimulusSettings.add_argument('-maxf', '--maxFrequency', type=float, default=36.0, help='The highest frequency value to measure (cycles per degree)')
	stimulusSettings.add_argument('-fr', '--frequencyResolution', type=int, default=20, help='The number of frequency steps')

	stimulusSettings.add_argument('--size', type=int, default=3, help='Gabor patch size in (degrees)')
	stimulusSettings.add_argument('--orientation', type=float, help='Orientation of gabor patch (degrees). If unspecified, each trial will be random')

	settings = argparseqt.groupingTools.parseIntoGroups(parser)
	if None in [settings['sessionID'], settings['distance_mm']]:
		settings = ui.getSettings(parser, settings, ['sessionID', 'distance_mm'])

	return settings

def main():
	from . import log
	settings = getSettings()

	if not settings is None:
		logPath = pathlib.Path(settings['outputFile']).parent
		log.startLog(settings['sessionID'], logPath)
		run(settings)

if __name__ == '__main__':
	main()

QuickCSF/screens.py
# -*- coding: utf-8 -*
'''Useful functions for working with different monitors'''

import math

from qtpy import QtCore, QtGui


def getPrimaryScreen():
	return QtGui.QGuiApplication.primaryScreen()


def getSecondaryScreen():
	primary = getPrimaryScreen()
	for screen in QtGui.QGuiApplication.screens():
		if screen.name() != primary.name():
			return screen


def getActiveScreen(widget=None):
	'''Returns the screen object which is currently active

	Args:
		widget (optional): if specified, it's center will be used to determine the "active" screen.
			If not specified, widget with focus will be used
			If no widget has focus, returns primary screen
	'''
	if widget is None:
		widget = QtGui.QGuiApplication.focusWindow()
		if widget is None:
			return getPrimaryScreen()

	# windowCenter = widget.geometry().center()
	# return QtGui.QGuiApplication.screenAt(windowCenter)
	return QtGui.QGuiApplication.primaryScreen()  # qt 5.9 compatability


def degreesToPixels(degrees, distance_mm, screen=None):
	'''Converts visual angle to pixels
		Note:
			Does NOT account for monitor flatness!

		Args:
			degrees: the size of the visual angle
			distance_mm: the distance between the observer and the screen in millimeters
			screen (optional): which screen's metrics to use
	'''
	if screen is None:
		screen = getActiveScreen()

	return degrees / math.degrees(math.atan((screen.physicalSize().width() / screen.geometry().width()) / distance_mm))


def moveToScreen(window, screen):
	'''Moves a window to a specific screen and attempts to preserve state / location

		Args:
			window: the window to be moved
			screen: the target screen to receive the window
	'''
	state = window.windowState()
	window.showNormal()
	window.move(screen.geometry().center() - QtCore.QPoint(window.geometry().width(), window.geometry().height()) / 2)
	QtCore.QTimer.singleShot(1, lambda: window.setWindowState(state))


def moveToSecondaryScreen(window):
	moveToScreen(window, getSecondaryScreen())


def moveToPrimaryScreen(window):
	moveToScreen(window, getPrimaryScreen())


if __name__ == '__main__':
	from qtpy import QtWidgets
	import sys

	def updateButton():
		screen = getActiveScreen()
		print(screen.name())

		window.setText(str(degreesToPixels(3, 750)))
		# moveToSecondaryScreen(window)

	app = QtWidgets.QApplication(sys.argv)
	window = QtWidgets.QPushButton()
	window.setText('hi')

	window.clicked.connect(updateButton)

	window.show()
	app.exec_()


QuickCSF/simulate.py
# -*- coding: utf-8 -*
'''Simulate a QuickCSF experiment'''

import logging
import argparse
import time
import math

import numpy

import matplotlib
import matplotlib.pyplot as plt
import pathlib
import argparseqt.gui
import argparseqt.groupingTools

from . import QuickCSF

logger = logging.getLogger('QuickCSF.simulate')

def plot(qCSFEstimator, graph=None, unmappedTrueParams=None, showNumbers=True):
	'''Generate a plot of estimates from QuickCSF, along with history of responses and true parameter values'''

	if graph is None:
		fig = plt.figure()
		graph = fig.add_subplot(1, 1, 1)

		plt.ion()
		plt.show()

	frequencyDomain = QuickCSF.makeFrequencySpace(.005, 80, 50).reshape(-1,1)

	if unmappedTrueParams is not None:
		truthData = QuickCSF.csf_unmapped(unmappedTrueParams.reshape(1, -1), frequencyDomain)
		truthData = numpy.power(10, truthData)
		truthLine = graph.fill_between(
			frequencyDomain.reshape(-1),
			truthData.reshape(-1),
			color=(1, 0, 0, .5)
		)
	else:
		truthData = None

	estimatedParamMeans = qCSFEstimator.getResults(leaveAsIndices=True)
	estimatedData = QuickCSF.csf_unmapped(estimatedParamMeans.reshape(1, -1), frequencyDomain)
	estimatedData = numpy.power(10, estimatedData)

	estimatedLine = graph.fill_between(
		frequencyDomain.reshape(-1),
		estimatedData.reshape(-1),
		color=(0, 0, 1, .4)
	)

	## Chart responses
	positives = {'f':[], 's':[]}
	negatives = {'f':[], 's':[]}
	for record in qCSFEstimator.responseHistory:
		stimValues = record[0]
		targetArray = positives if record[1] else negatives
		targetArray['f'].append(stimValues[1])
		targetArray['s'].append(1/stimValues[0])

	graph.plot(positives['f'], positives['s'], 'o', markersize=4, color=(.2, 1, .2))
	graph.plot(negatives['f'], negatives['s'], 'x', markersize=5, color=(1,0,0), markeredgewidth=2)

	graph.set_xlabel('Spatial frequency (CPD)')
	graph.set_xscale('log')
	graph.set_xlim((.25, 64))
	graph.set_xticks([1, 2, 4, 8, 16, 32])
	graph.get_xaxis().set_major_formatter(matplotlib.ticker.ScalarFormatter())

	graph.set_ylabel('Sensitivity (1/contrast)')
	graph.set_yscale('log')
	graph.set_ylim((1, 400))
	graph.set_yticks([2, 10, 50, 200])
	graph.get_yaxis().set_major_formatter(matplotlib.ticker.ScalarFormatter())

	graph.grid()

	if showNumbers:
		estimatedParamMeans = QuickCSF.mapCSFParams(estimatedParamMeans, exponify=True)
		estimatedParamMeans = estimatedParamMeans.reshape(1,-1).tolist()[0]
		paramEstimates = '%03.2f, %.4f, %.4f, %.4f' % tuple(estimatedParamMeans)
		estimatedLine.set_label(f'Estim: {paramEstimates}')

		if truthData is not None:
			trueParams = QuickCSF.mapCSFParams(unmappedTrueParams, True).T.tolist()[0]
			trueParams = '%03.2f, %.4f, %.4f, %.4f' % tuple(trueParams)
			truthLine.set_label(f'Truth: {trueParams}')

		graph.legend()

	plt.pause(0.001) # necessary for non-blocking graphing

	return graph

def runSimulation(
	trials=30,
	imagePath=None,
	usePerfectResponses=False,
	stimuli={
		'minContrast':0.01, 'maxContrast':1, 'contrastResolution':24,
		'minFrequency':.2, 'maxFrequency':36, 'frequencyResolution':20,
	},
	parameters={
		'truePeakSensitivity':18, 'truePeakFrequency':11,
		'trueBandwidth':12, 'trueBandwidth':11,
	},
):
	logger.info('Starting simulation')

	numpy.random.seed()

	if imagePath is not None:
		pathlib.Path(imagePath).mkdir(parents=True, exist_ok=True)

	stimulusSpace = numpy.array([
		QuickCSF.makeContrastSpace(stimuli['minContrast'], stimuli['maxContrast'], stimuli['contrastResolution']),
		QuickCSF.makeFrequencySpace(stimuli['minFrequency'], stimuli['maxFrequency'], stimuli['frequencyResolution'])
	])

	unmappedTrueParams = numpy.array([[
		parameters['truePeakSensitivity'],
		parameters['truePeakFrequency'],
		parameters['trueBandwidth'],
		parameters['trueBandwidth'],
	]])
	qcsf = QuickCSF.QuickCSFEstimator(stimulusSpace)

	graph = plot(qcsf, unmappedTrueParams=unmappedTrueParams)

	# Trial loop
	for i in range(trials):
		# Get the next stimulus
		stimulus = qcsf.next()
		newStimValues = numpy.array([[stimulus.contrast, stimulus.frequency]])

		# Simulate a response
		if usePerfectResponses:
			logger.debug('Simulating perfect response')
			frequency = newStimValues[:,1]
			trueSens = numpy.power(10, QuickCSF.csf_unmapped(unmappedTrueParams, numpy.array([frequency])))
			testContrast = newStimValues[:,0]
			testSens = 1 / testContrast

			response = trueSens > testSens
		else:
			logger.debug('Simulating human response response')
			p = qcsf._pmeas(unmappedTrueParams)
			response = numpy.random.rand() < p

		qcsf.markResponse(response)

		# Update the plot
		graph.clear()
		graph.set_title(f'Estimated Contrast Sensitivity Function ({i+1})')
		plot(qcsf, graph, unmappedTrueParams)

		if imagePath is not None:
			plt.savefig(pathlib.Path(imagePath+'/%f.png' % time.time()).resolve())

	logger.info('Simulation complete')
	print('******* History *******')
	for record in qcsf.responseHistory:
		print(f'\tf={record[0][1]},c={record[0][0]},r={record[1]}')

	print('***********************')

	paramEstimates = qcsf.getResults()
	logger.info('Results: ' + str(paramEstimates))

	trueParams = QuickCSF.mapCSFParams(unmappedTrueParams, True).T
	print('******* Results *******')
	print(f'\tEstimates = {paramEstimates}')
	print(f'\tActuals = {trueParams}')
	print('***********************')
	plt.ioff()
	plt.show()

def entropyPlot(qcsf):
	params = numpy.arange(qcsf.paramComboCount).reshape(-1, 1)
	stims = numpy.arange(qcsf.stimComboCount).reshape(-1,1)

	p = qcsf._pmeas(params, stims)

	pbar = sum(p)/len(params)
	hbar = sum(QuickCSF.entropy(p))/len(params)
	gain = QuickCSF.entropy(pbar)-hbar


	gain = -gain.reshape(qcsf.stimulusRanges[::-1]).T

	fig = plt.figure()
	graph = fig.add_subplot(1, 1, 1)
	plt.imshow(gain, cmap='hot')

	plt.ioff()
	plt.show()


if __name__ == '__main__':
	from . import log
	log.startLog()

	parser = argparse.ArgumentParser()

	parser.add_argument('-n', '--trials', type=int, help='Number of trials to simulate')
	parser.add_argument('--imagePath', default=None, help='If specified, path to save images')
	parser.add_argument('-perfect', '--usePerfectResponses', default=False, action='store_true', help='Whether to simulate perfect responses, rather than probablistic ones')

	stimuliSettings = parser.add_argument_group('stimuli')
	stimuliSettings.add_argument('-minc', '--minContrast', type=float, default=.01, help='The lowest contrast value to measure (0.0-1.0)')
	stimuliSettings.add_argument('-maxc', '--maxContrast', type=float, default=1.0, help='The highest contrast value to measure (0.0-1.0)')
	stimuliSettings.add_argument('-cr', '--contrastResolution', type=int, default=24, help='The number of contrast steps')

	stimuliSettings.add_argument('-minf', '--minFrequency', type=float, default=0.2, help='The lowest frequency value to measure (cycles per degree)')
	stimuliSettings.add_argument('-maxf', '--maxFrequency', type=float, default=36.0, help='The highest frequency value to measure (cycles per degree)')
	stimuliSettings.add_argument('-fr', '--frequencyResolution', type=int, default=20, help='The number of frequency steps')

	parameterSettings = parser.add_argument_group('parameters')
	parameterSettings.add_argument('-s', '--truePeakSensitivity', type=int, default=18, help='True peak sensitivity (index)')
	parameterSettings.add_argument('-f', '--truePeakFrequency', type=int, default=11, help='True peak frequency (index)')
	parameterSettings.add_argument('-b', '--trueBandwidth', type=int, default=12, help='True bandwidth (index)')
	parameterSettings.add_argument('-d', '--trueDelta', type=int, default=11, help='True delta truncation (index)')

	settings = argparseqt.groupingTools.parseIntoGroups(parser)
	if settings['trials'] is None:
		from qtpy import QtWidgets
		from . import ui
		app = QtWidgets.QApplication(sys.argv)
		settings = ui.getSettings(parser, settings, ['trials'])

	if settings is not None:
		runSimulation(**settings)
